cmd/eventkitd-sink-bigquery/main.go

```diff
@@ -0,0 +1,207 @@
+
+	for _, event := range packet.Packet.Events {
+		for _, scope := range event.Scope {
+			k := scope + "." + event.Name
+
+			records[k] = append(records[k], &Record{
+				Application: Application{
+					Name:     packet.Packet.Application,
+					Instance: packet.Packet.Instance,
+					Version:  packet.Packet.ApplicationVersion,
+				},
+				StartTime:  packet.Packet.StartTimestamp.AsTime(),
+				ReceivedAt: packet.ReceivedAt,
+				SendOffset: packet.Packet.SendOffset.AsDuration(),
+				Source:     packet.Source,
+				Tags:       event.Tags,
+			})
+		}
+	}
+
+	for table, events := range records {
+		err := b.dataset.Table(table).Inserter().Put(ctx, events)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type Config struct {
+	Address *string
+	Workers *int
+
+	Google struct {
+		ProjectID *string
+
+		BigQuery struct {
+			Dataset *string
+		}
+	}
+}
+
+type Packet struct {
+	Packet     *pb.Packet
+	Source     *net.UDPAddr
+	ReceivedAt time.Time
+}
+
+func main() {
+	log, _ := zap.NewProduction()
+
+	cfg := Config{}
+	cfg.Address = flag.String("bind-address", ":9002", "udp address to listen on")
+	cfg.Workers = flag.Int("workers", runtime.NumCPU(), "number of workers")
+	cfg.Google.ProjectID = flag.String("google-project-id", os.Getenv("GOOGLE_PROJECT_ID"), "configure which google project is being used (env: GOOGLE_PROJECT_ID)")
+	cfg.Google.BigQuery.Dataset = flag.String("google-bigquery-dataset", os.Getenv("GOOGLE_BIGQUERY_DATASET"), "configure which dataset is being used")
+	flag.Parse()
+
+	ctx, done := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer done()
+
+	client, err := bigquery.NewClient(ctx, *cfg.Google.ProjectID)
+	if err != nil {
+		log.Error("bigquery client connection failed", zap.Error(err))
+		os.Exit(1)
+		return
+	}
+
+	log.Info("starting server", zap.Stringp("address", cfg.Address))
+	listener, err := transport.ListenUDP(*cfg.Address)
+	if err != nil {
+		log.Error("failed to bind to udp address")
+		os.Exit(1)
+		return
+	}
+
+	sink := &BigQuerySink{dataset: client.Dataset(*cfg.Google.BigQuery.Dataset)}
+
+	group, ctx := errgroup.WithContext(ctx)
+	workQueue := make(chan *Packet, *cfg.Workers*2)
+
+	for i := 0; i < *cfg.Workers; i++ {
+		group.Go(func() error {
+			for {
+				select {
+				case <-ctx.Done():
+					return nil
+				case packet := <-workQueue:
+					err = sink.Receive(ctx, packet)
+					if err != nil {
+						log.Error("failed to process packet", zap.Error(err))
+					}
+				}
+			}
+		})
+	}
+
+	noWait := &errgroup.Group{}
+	noWait.Go(func() error {
+		t := time.NewTicker(200 * time.Millisecond)
+		defer func() { t.Stop() }()
+
+		for {
+			packet, source, err := listener.Next()
+			if err != nil {
+				log.Error("failed to read udp packet", zap.Error(err))
+				continue
+			}
+
+			workQueue <- &Packet{
+				Packet:     packet,
+				Source:     source,
+				ReceivedAt: time.Now(),
+			}
+
+			t = time.NewTicker(200 * time.Millisecond)
```

Comment:
what's the ticker for? it doesn't look like it's being used

Reply:
woops, had it in there and forgot to clean it up.